server: proper transaction state management in sql-over-http

ajwerner opened the pull request:
First 4 commits are https://github.com/cockroachdb/cockroach/pull/86427.
Next commit is https://github.com/cockroachdb/cockroach/pull/86461.

We need to construct the internal executor in the context of the transaction
so that we can make sure that its side-effects are properly managed. Without
this change, we'd be throwing away all of the extraTxnState between each
statement. We'd fail to create the jobs (which we defer to the end of the
transaction), and we'd fail to run those jobs and check for errors. We'd
also fail to validate the two-version invariant or wait for one version.

Fixes https://github.com/cockroachdb/cockroach/issues/86332

Release justification: Fixes critical bugs in new functionality.

Release note: None

ajwerner committed:
server: proper transaction state management in sql-over-http

We need to construct the internal executor in the context of the transaction
so that we can make sure that its side-effects are properly managed. Without
this change, we'd be throwing away all of the extraTxnState between each
statement. We'd fail to create the jobs (which we defer to the end of the
transaction), and we'd fail to run those jobs and check for errors. We'd
also fail to validate the two-version invariant or wait for one version.

Fixes #86332

Release justification: Fixes critical bugs in new functionality.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86433)
<!-- Reviewable:end -->

ajwerner commented:
@rafiss or @knz can you review the last commit on this one (and feel empowered to review https://github.com/cockroachdb/cockroach/pull/86427 in the process)?

ajwerner committed:
sql/catalog/descs: make it possible to use a Stepping Txn

This is needed both to ensure that the transactions which
purposefully use the internal executor can avoid the halloween
problem and so that we can use a default QoS level for those
transaction.

Release justification: part of a bug fix

Release note: None

knz reviewed (commented):
Last commit LGTM.

I did not review the others.

> Reviewed 3 of 3 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86433#-:-N9rAktv2fa6as4hx30a:b-cj5kwh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)



<!-- Sent from Reviewable.io -->

ajwerner commented:
TFTR!

bors r+
